File: Backend/models/fileCache.js
const db = require('../config/db');
const USE_POSTGRES = process.env.USE_POSTGRES === 'true';

/**
 * File Cache Model
 * Handles caching of file processing results
 * Works with both SQLite and PostgreSQL
 */
class FileCache {
    /**
     * Find cached file by hash and version
     * @param {string} hash - SHA-256 hash of file
     * @param {number} version - Processing version
     * @returns {Promise<object|null>} - Cached data or null
     */
    static async findByHash(hash, version) {
        if (USE_POSTGRES) {
            // PostgreSQL async query
            const result = await db.query(
                `SELECT * FROM file_cache 
                 WHERE file_hash = $1 AND processing_version = $2`,
                [hash, version]
            );
            
            if (result.rows.length > 0) {
                // Update last accessed time
                await this.updateLastAccessed(hash, version);
                
                const row = result.rows[0];
                // Parse JSON fields (PostgreSQL returns JSONB as objects)
                return {
                    ...row,
                    departmental_summaries: typeof row.departmental_summaries === 'string' 
                        ? JSON.parse(row.departmental_summaries) 
                        : row.departmental_summaries,
                    metadata: row.metadata ? (
                        typeof row.metadata === 'string' 
                            ? JSON.parse(row.metadata) 
                            : row.metadata
                    ) : null
                };
            }
            
            return null;
        } else {
            // SQLite sync query
            const stmt = db.prepare(`
                SELECT * FROM file_cache 
                WHERE file_hash = ? AND processing_version = ?
            `);
            
            const result = stmt.get(hash, version);
            
            if (result) {
                // Update last accessed time
                this.updateLastAccessed(hash, version);
                
                // Parse JSON fields
                return {
                    ...result,
                    departmental_summaries: JSON.parse(result.departmental_summaries),
                    metadata: result.metadata ? JSON.parse(result.metadata) : null
                };
            }
            
            return null;
        }
    }

    /**
     * Create new cache entry
     * @param {object} data - Cache data
     * @returns {Promise<object>} - Created entry
     */
    static async create(data) {
        const { fileHash, processingVersion, originalFilename, extractedText, departmentalSummaries, metadata } = data;
        
        if (USE_POSTGRES) {
            // PostgreSQL async query
            const result = await db.query(
                `INSERT INTO file_cache (
                    file_hash,
                    processing_version,
                    original_filename, 
                    extracted_text, 
                    departmental_summaries, 
                    metadata
                ) VALUES ($1, $2, $3, $4, $5, $6)
                RETURNING *`,
                [
                    fileHash,
                    processingVersion,
                    originalFilename,
                    extractedText,
                    JSON.stringify(departmentalSummaries),
                    metadata ? JSON.stringify(metadata) : null
                ]
            );
            
            console.log(`✅ Cached file: ${originalFilename} (hash: ${fileHash.substring(0, 8)}..., v${processingVersion})`);
            
            return { id: result.rows[0].id, ...data };
        } else {
            // SQLite sync query
            const stmt = db.prepare(`
                INSERT INTO file_cache (
                    file_hash,
                    processing_version,
                    original_filename, 
                    extracted_text, 
                    departmental_summaries, 
                    metadata
                ) VALUES (?, ?, ?, ?, ?, ?)
            `);
            
            const result = stmt.run(
                fileHash,
                processingVersion,
                originalFilename,
                extractedText,
                JSON.stringify(departmentalSummaries),
                metadata ? JSON.stringify(metadata) : null
            );
            
            console.log(`✅ Cached file: ${originalFilename} (hash: ${fileHash.substring(0, 8)}..., v${processingVersion})`);
            
            return { id: result.lastInsertRowid, ...data };
        }
    }

    /**
     * Update last accessed timestamp
     * @param {string} hash - File hash
     * @param {number} version - Processing version
     */
    static async updateLastAccessed(hash, version) {
        if (USE_POSTGRES) {
            // PostgreSQL async query
            await db.query(
                `UPDATE file_cache 
                 SET last_accessed_at = CURRENT_TIMESTAMP 
                 WHERE file_hash = $1 AND processing_version = $2`,
                [hash, version]
            );
        } else {
            // SQLite sync query
            const stmt = db.prepare(`
                UPDATE file_cache 
                SET last_accessed_at = CURRENT_TIMESTAMP 
                WHERE file_hash = ? AND processing_version = ?
            `);
            
            stmt.run(hash, version);
        }
    }

    /**
     * Get cache statistics
     * @returns {Promise<object>} - Cache stats
     */
    static async getStats() {
        if (USE_POSTGRES) {
            // PostgreSQL async queries
            const countResult = await db.query('SELECT COUNT(*) as count FROM file_cache');
            const sizeResult = await db.query('SELECT SUM(LENGTH(extracted_text::text)) as size FROM file_cache');
            
            const count = parseInt(countResult.rows[0].count);
            const size = parseInt(sizeResult.rows[0].size || 0);
            
            return {
                totalEntries: count,
                totalSize: size,
                totalSizeMB: (size / (1024 * 1024)).toFixed(2)
            };
        } else {
            // SQLite sync queries
            const countStmt = db.prepare('SELECT COUNT(*) as count FROM file_cache');
            const sizeStmt = db.prepare('SELECT SUM(LENGTH(extracted_text)) as size FROM file_cache');
            
            const count = countStmt.get().count;
            const size = sizeStmt.get().size || 0;
            
            return {
                totalEntries: count,
                totalSize: size,
                totalSizeMB: (size / (1024 * 1024)).toFixed(2)
            };
        }
    }
}

module.exports = FileCache;

File: Backend_py/models/file_cache.py
"""
File Cache Model - PostgreSQL Implementation
Handles file caching operations
"""
import json
import logging
from typing import Optional, Dict, Any
from datetime import datetime
from core.sqlalchemy_db import get_db_session
from models.sqlalchemy_models import FileCache as FileCacheModel

logger = logging.getLogger(__name__)

class FileCache:
    @staticmethod
    def find_by_hash(file_hash: str, version: int) -> Optional[Dict[str, Any]]:
        """Find cached file by hash and version"""
        db = get_db_session()
        try:
            cached_file = db.query(FileCacheModel).filter(
                FileCacheModel.file_hash == file_hash,
                FileCacheModel.processing_version == version
            ).first()
            
            if cached_file:
                # Update last accessed time
                cached_file.last_accessed_at = datetime.utcnow()
                db.commit()
                
                result = {
                    "id": cached_file.id,
                    "file_hash": cached_file.file_hash,
                    "processing_version": cached_file.processing_version,
                    "original_filename": cached_file.original_filename,
                    "extracted_text": cached_file.extracted_text,
                    "departmental_summaries": cached_file.departmental_summaries or {},
                    "metadata": cached_file.cache_metadata or {},
                    "created_at": cached_file.created_at,
                    "last_accessed_at": cached_file.last_accessed_at
                }
                
                return result
            return None
        except Exception as e:
            logger.error(f"Error finding in cache: {str(e)}")
            return None
        finally:
            db.close()

    @staticmethod
    def create(data: Dict[str, Any]):
        """Create or update a cached file"""
        db = get_db_session()
        try:
            file_hash = data["fileHash"]
            version = data["processingVersion"]
            filename = data["originalFilename"]
            text = data["extractedText"]
            summaries = data["departmentalSummaries"]
            metadata = data.get("metadata")
            
            # Try to find existing cache entry
            cached_file = db.query(FileCacheModel).filter(
                FileCacheModel.file_hash == file_hash,
                FileCacheModel.processing_version == version
            ).first()
            
            if cached_file:
                # Update existing entry
                cached_file.original_filename = filename
                cached_file.extracted_text = text
                cached_file.departmental_summaries = summaries
                cached_file.cache_metadata = metadata
                cached_file.last_accessed_at = datetime.utcnow()
            else:
                # Create new entry
                cached_file = FileCacheModel(
                    file_hash=file_hash,
                    processing_version=version,
                    original_filename=filename,
                    extracted_text=text,
                    departmental_summaries=summaries,
                    cache_metadata=metadata,
                    last_accessed_at=datetime.utcnow()
                )
                db.add(cached_file)
            
            db.commit()
            logger.info(f"✅ Cached file: {filename} (hash: {file_hash[:8]}..., v{version})")
        except Exception as e:
            db.rollback()
            logger.error(f"Error creating cache: {str(e)}")
        finally:
            db.close()
